core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -207,13 +243,155 @@ Map<TopicIdPartition, FetchRequest.PartitionData> acquirablePartitions() {
+            }
+            sharePartition.updateFetchOffsetMetadata(Optional.of(replicaManagerLogReadResult.info().fetchOffsetMetadata));
+        }
+        return replicaManagerReadResponseData;
+    }
+
+    private boolean isMinBytesSatisfied(Map<TopicIdPartition, FetchRequest.PartitionData> topicPartitionData) {
+        long accumulatedSize = 0;
+        for (Map.Entry<TopicIdPartition, FetchRequest.PartitionData> entry : topicPartitionData.entrySet()) {
+            TopicIdPartition topicIdPartition = entry.getKey();
+            FetchRequest.PartitionData partitionData = entry.getValue();
+            LogOffsetMetadata endOffsetMetadata = endOffsetMetadataForTopicPartition(topicIdPartition);
+
+            if (endOffsetMetadata == LogOffsetMetadata.UNKNOWN_OFFSET_METADATA)
+                continue;
+
+            SharePartition sharePartition = sharePartitions.get(topicIdPartition);
+
+            Optional<LogOffsetMetadata> optionalFetchOffsetMetadata = sharePartition.fetchOffsetMetadata();
+            if (optionalFetchOffsetMetadata.isEmpty() || optionalFetchOffsetMetadata.get() == LogOffsetMetadata.UNKNOWN_OFFSET_METADATA)
+                continue;
+            LogOffsetMetadata fetchOffsetMetadata = optionalFetchOffsetMetadata.get();
+
+            if (fetchOffsetMetadata.messageOffset > endOffsetMetadata.messageOffset) {
+                log.debug("Satisfying delayed share fetch request for group {}, member {} since it is fetching later segments of " +
+                    "topicIdPartition {}", shareFetchData.groupId(), shareFetchData.memberId(), topicIdPartition);
+                return true;
+            } else if (fetchOffsetMetadata.messageOffset < endOffsetMetadata.messageOffset) {
+                if (fetchOffsetMetadata.onOlderSegment(endOffsetMetadata)) {
+                    // This can happen when the fetch operation is falling behind the current segment or the partition
+                    // has just rolled a new segment.
+                    log.debug("Satisfying delayed share fetch request for group {}, member {} immediately since it is fetching older " +
+                        "segments of topicIdPartition {}", shareFetchData.groupId(), shareFetchData.memberId(), topicIdPartition);
+                    return true;
+                } else if (fetchOffsetMetadata.onSameSegment(endOffsetMetadata)) {
+                    // we take the partition fetch size as upper bound when accumulating the bytes.
+                    long bytesAvailable = Math.min(endOffsetMetadata.positionDiff(fetchOffsetMetadata), partitionData.maxBytes);
+                    accumulatedSize += bytesAvailable;
+                }
+            }
+        }
+        return accumulatedSize >= shareFetchData.fetchParams().minBytes;
+    }
+
+    private LogOffsetMetadata endOffsetMetadataForTopicPartition(TopicIdPartition topicIdPartition) {
+        Partition partition = replicaManager.getPartitionOrException(topicIdPartition.topicPartition());
```

Comment:
What @junrao meant was that it should not be a top level rather partition level error.

Reply:
hey @apoorvmittal10 , we are doing partition level error handling in `onComplete` using `sharePartitionManager.handleFetchException`. When we get an exception in this line `replicaManager.getPartitionOrException`, we directly call `forceComplete` (mentioned in above comments), and that does a partition level handling.

Reply:
It's missing yet, you can log a jira for me to fix as I am doing handling anyways.

Reply:
logged a JIRA https://issues.apache.org/jira/browse/KAFKA-17943 for the same.